---
title: "DNA Variant Analysis with R Bioconductor" 
authors: "Karla Lindquist"
output: html_notebook
---

------------------------------------------------------------------------

#### In this workshop, our objectives are to:

1.  Download RNA-seq files from GEO
2.  Annotate the samples
3.  Calculate expression as read counts/million (CPM) and log2(CPM)
4.  Filter out genes with low expression and plot
5.  Normalize the filtered expression data and plot
6.  Perform unsupervised clustering of expression data
7.  Perform differential expression analysis
8.  Create heatmaps and volcano plots
9.  Perform enrichment and pathway analysis

------------------------------------------------------------------------

#### Install packages

Instructions and notes:

-   Please make sure that you have *current* versions of R, Bioconductor, and RStudio installed.

    -   [R](https://cran.r-project.org/)

    -   [RStudio (free desktop version)](https://www.rstudio.com/products/rstudio/download/)

    -   [Bioconductor](https://bioconductor.org/install/)\

-   I recommend you start a fresh R/RStudio session *before* running the code below. If asked to restart R again during the installation process, it means you already have the package so you can say No.

-   If asked "Update all/some/none?", select No [n] during class (other times select Yes [y]). Older versions of the packages that we will be using will probably work OK.

-   If asked "Do you want to install from sources the package which needs compilation?", select No [n] during class (other times select Yes [y]). Note that source versions are typically a little newer than binary versions.

-   Lazy loading errors can be ignored, as can warnings about the version of R packages were built under.

-   If you get a message saying "Error in install.packages : Updating loaded packages" do not worry about this.

-   We will not be re-installing the packages during class, so it is important that you make sure that you have these installed before class.

-   If you have questions about installation issues, feel free to email [karla.lindquist\@ucsf.edu](mailto:karla.lindquist@ucsf.edu){.email}.

```{r message=FALSE}
## install Bioconductor
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(version = "3.15")  

## list Bioconductor packages
bpkgs <- c("GEOquery","org.Mm.eg.db","GenomeInfoDbData","TxDb.Mmusculus.UCSC.mm10.knownGene","Mus.musculus","BiocStyle","limma","edgeR","RColorBrewer","gplots","gage","clusterProfiler","ReactomePA","GOSemSim","DOSE","enrichplot")
## note: some may not be necessary to follow along in the workshop (more for demo purposes)

## install Bioconductor packages
BiocManager::install(bpkgs, force = TRUE) 


## list CRAN packages
rpkgs <- c("ggplot2","here","cli","R.utils","rmarkdown","knitr","utf8") 
## note: some but not all may need to install utf8 to get all Bioconductor packages to install; rmarkdown and knitr are only needed if you are using R notebooks

## install CRAN packages
install.packages(rpkgs)
```

#### Load all packages

```{r message=FALSE}
## list and load all packages
pkgs <- c(bpkgs, rpkgs)
for(package in pkgs) {
  do.call("library", list(package))
}
```

------------------------------------------------------------------------

#### Acknowledgments

To find package citations, you can use the `citation()` function with package name as the argument.

```{r}
citation("enrichplot")
```
