---
title: "Annotate the samples" 
authors: "Karla Lindquist"
date: "7/26/2021"
output: html_notebook
---

#### *Objective 2. Annotate the samples*

Packages used in this notebook: `edgeR`, `org.Mm.eg.db`. First verify that these are installed and loaded.

------------------------------------------------------------------------

First, let's read in all of the files that we left out the first time around. In Objective 1 we created a vector of filenames stored in the `files` object. It is assumed that this object is still in your environment. If it is not found then you need to re-create it.

```{r}
files
```

The `readDGE()` function from the `edgeR` package conveniently reads all of the files at once. Let's take another look at this object that we named simply `dge`. Re can re-run the code to recreate this.

```{r}
dge <- readDGE(files=files, path=paste0(projdir, "data"), columns=c(1,3))

dge
```

------------------------------------------------------------------------

##### Label the sample groups

Now we need to indicate what the different sample groups are. In this case we'll be able to identify these groups by looking at the Samples section in GEO ([GSE63310](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE63310)). In our own experimental data, we'd already have this information by virtue of designing the experiment ourselves.

In this case we have 3 groups: LP (luminal progenitor), ML (mature luminal), and Basal epithelial cells.

Let's add the sample annotations to the **group** column of the DGEList object `dge`.

```{r}
dge$samples$group <- as.factor(c("LP", "ML", "Basal", "Basal", "ML", "LP", "Basal", "ML", "LP"))
dge$samples
```

##### Label the gene names

Now let's look at the raw counts associated with each gene. The counts part of the DGEList contains gene names (ENTREZIDs) in rows and sample names in columns. You can use the `rownames` function on this object to create a vector of all gene names.

```{r}
head(dge$counts)
dge_ids <- rownames(dge)

head(dge_ids) ## look at first 6 gene names

length(dge_ids) ## find out how many genes are represented 
```

Now say we want to get gene SYMBOLS as well - we need this information from the mouse genome database `org.Mm.eg.db`. We're able to do this by using the mouse genome info this is built into the `AnnotationDbi` package that comes with Bioconductor. Note the use of `keys=dge_ids` here (we wouldn't want to type them all out since this is a long vector). We will call the new vector with SYMBOLS `gene_ids`.

```{r}
?AnnotationDbi::select
gene_ids <- AnnotationDbi::select(org.Mm.eg.db, 
    keys=dge_ids, 
    keytype="ENTREZID",
    columns=c("SYMBOL"))
head(gene_ids)
```

Note: 'select()' returned 1:1 mapping between keys and columns is what we are expecting here - there should be a 1:1 mapping between ENTREZIDs and gene SYMBOLS!

Question: Can you think of situations where you might see a 1:many mapping (hint: you can map between genes and other types of features that go along with them other than their names).

------------------------------------------------------------------------

Now we can add these identifiers to the DGEList object in addition to the samples and counts.

```{r}
names(dge)
```

```{r}
dge$gene_ids <- gene_ids
names(dge)
```
