---
title: "Annotate the samples" 
author: "Karla Lindquist"
output: html_notebook
---

#### *Objective 2. Annotate the samples*

Packages used in this notebook: `edgeR`, `org.Mm.eg.db`. Make sure these packages are installed and loaded.

------------------------------------------------------------------------

As seen in the previous notebook, the `readDGE()` function from the `edgeR` package conveniently reads all of the files at once. Let's take another look at this object that we named `dge`.

```{r}
dge
```

You can also look at the first few rows of the raw counts which is in a matrix, so you can access the values using the `$` operator.

```{r}
names(dge)
class(dge$counts)
head(dge$counts)
```

------------------------------------------------------------------------

##### Label the gene names

The counts part of the DGE object contains gene names as ENTREZIDs in rows and sample or file names in columns. You may want to convert these ENTREZIDs to gene symbols instead, which are easier to read. You can use the `rownames` function on the `dge` object to create a vector of ENTREZIDs that we can then translate to symbols.

```{r}
dge_ids <- rownames(dge)

head(dge_ids) ## look at first 6 gene names

length(dge_ids) ## find out how many genes are represented 
```

To annotate the data with gene SYMBOLS instead of ENTREZIDs, we need to use the mouse genome database `org.Mm.eg.db` since our samples come from mice.

Let\'s use the `select()` function from the `AnnotationDbi` package (comes with Bioconductor). Note that there are multiple packages that support a function with the same name. We want to look at the one from the package, so we will call up the help file from that one with `?AnnotationDbi::select`. Take some time to understand this function - it is used a lot with Bioconductor packages.

Note the use of `keys = dge_ids` here (we wouldn't want to type them all out since this is a long vector). We will create a new vector but using SYMBOLS instead of ENTREZIDs, and we'll call this `gene_ids`.

```{r}
?AnnotationDbi::select
gene_ids <- AnnotationDbi::select(org.Mm.eg.db, 
    keys = dge_ids, 
    keytype = "ENTREZID",
    columns = "SYMBOL")
head(gene_ids)
```

Note: `select()` returned 1:1 mapping between keys and columns is what we are expecting here - there should be a 1:1 mapping between ENTREZIDs and gene SYMBOLs, so this is good. If you get 1:many or many:1 then you want to make sure you understand why ... e.g. this can happen if you are mapping genes names to multiple transcripts.

------------------------------------------------------------------------

Now, we can add these identifiers to our `dge` count matrix by re-assigning the row names with SYMBOLs from our `gene_ids` object.

```{r}
head(dge$counts)
head(gene_ids)
```

```{r}
rownames(dge$counts) <- gene_ids$SYMBOL
dge
```
