---
title: "Estimate expression levels" 
author: "Karla Lindquist"
output: html_notebook
---

#### *Objective 3. Calculate expression as read counts/million (CPM) and log2(CPM)*

Packages used in this notebook: `edgeR`. Make sure this package is installed and loaded.

------------------------------------------------------------------------

##### Comments on measures used to estimate expression levels from raw read counts

We will convert our raw counts into counts per million (CPM). We will also be calculating log2 counts/million, log2(CPM). This is related to expression. It isn't necessarily the best measurement to use but it's the simplest one. You want to become familiar with the different ways of measuring expression and choose the best one for your data.

Other measures to estimate expression from raw counts that you may hear about are RPKM and FPKM. For more discussion about this, see this article by Harold Pimentel called [What the FPKM? A review of RNA-Seq expression units](https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/). Here are a few excerpts from this article to summarize:

-   Counts per million (CPM) mapped reads are counts scaled by the number of fragments you sequenced (![N](https://s0.wp.com/latex.php?latex=N&bg=ffffff&fg=000000&s=0&c=20201002)) times one million.

-   Reads per kilobase of exon per million reads mapped (RPKM) or the more generic FPKM (substitute reads with fragments) are essentially the same thing. Contrary to some misconceptions, FPKM is not 2 \* RPKM if you have paired-end reads. FPKM == RPKM if you have single-end reads, and saying RPKM when you have paired-end reads is just weird, so don't do it :).

------------------------------------------------------------------------

##### Calculate counts per million (CPM) and log-transform

The `cpm()` function from the `edgeR` package calculates CPM values using the `dge` object that we have been working with. It can also calculate log2(CPM) values, which we will do by adding the log argument.

```{r}
?cpm
cpm <- cpm(dge)
log2_cpm <- cpm(dge, log=TRUE)
```

Take a look at raw counts, raw CPM, and log2(CPM) values.

```{r}
head(dge$counts) ## raw read counts
head(cpm) ## counts per million mapped reads
head(log2_cpm) ## log-transformed cpm
```

Note: when a raw count is zero, the `cpm()` function will apply a pseudocount to avoid talking the log of zero, which is not a number. We'll be filtering these out anyway.

Let's look at the distribution of each with simple histograms. The log transformation just makes the non-0 values more normally distributed which can simplify statistical testing of differences between sample types (more on this later).

```{r}
hist(cpm)
hist(log2_cpm)
```
