---
title: "Visualize differential results" 
author: "Karla Lindquist"
output: html_notebook
---

#### *Objective 8. Create heatmap and volcano plots*

Packages used in this notebook: `limma`, `gplots`, `RColorBrewer`. Make sure these packages are installed and loaded.

------------------------------------------------------------------------

Now that we've done some analysis of the data, we can visualize the results. While figures may not convey the depth of information that quantitative outputs might display, it does make our results far easier to interpret at a glance. Here we'll do a quick survey of some of the more popular RNA-seq visualizations.

------------------------------------------------------------------------

##### Hierarchical clustering with heatmaps

Hierarchical clustering can be a potent way to examine differences between samples, and heatmaps can succintly visualise the clustering of your samples. We can do this using the `heatmap.2()` function from the `gplots` package. In this example `heatmap.2()` calculates a matrix of euclidean distances from the log2(CPM ) values for the 500 most significant genes.

We can get the gene SYMBOLs for the 50 most significant genes by basically doing a sort by p-values and then select 1-50. You can sort by adjusted p-values for pairwise comparisons, or as I demonstrate here, you can sort by the F-statistic p-values which represent the overall significance of differences across the 3 sample groups. It is probably easier to convert your results in the `efit` object to a data frame first - let's call this `efit_df`.

```{r}
efit_df <- as.data.frame(efit, row.names = rownames(efit)) ## create data frame, retaining genes names as row names
efit_df <- efit_df[order(efit_df$F.p.value),] ## sort by F statistic p-values
head(efit_df)
```

If you want to see this full table more easily, you can show it in a new window using the tiny and faint icon at the top right corner of the output. Let's look at the top genes.

```{r}
top50 <- rownames(efit_df)[1:50]
top50
```

Now let's subset the `log2cpm_norm` matrix contain to contain just the top 50 genes using this vector of gene names. The heatmap needs to use these normalized expression levels to plot (more on this below).

```{r}
log2cpm_top50 <- log2cpm_norm[top50,]
dim(log2cpm_top50) ## should be 50 genes x 9 samples
head(log2cpm_top50)
```

Now plot the heatmap with these 50 genes. First let's look at the help file for the `heatmap.2` function. There are a lot of options! The easiest way to customize your own is to look at the examples and try a few different things, or look online to find some others that you like. We will make a pretty bare-bones heatmap here.

We will keep the samples together in columns and use the same colors that we used in the cluster plot in Objective 6 to indicate which is which. So at the top we will have the 3 ML samples (green), 3 LP samples (blue), and the 3 Basal samples (red).

The dendograms are based on hierarchical clustering to show similar samples. By default, `heatmap.2` uses Euclidean distances with complete agglomeration. Different heatmap functions will use different methods, and there are lots of tutorials out there that can help you understand the differences between them, e.g. see "[Heatmap in R: Static and Interactive Visualization](https://www.datanovia.com/en/lessons/heatmap-in-r-static-and-interactive-visualization/)" from Datanovia. If you get an error message saying "Error in plot.new() : figure margins too large" that just means that the plot can't fit on your screen. You can always try to comment the `png()` and `dev.off()` lines below and see if it works for you in the RStudio session. The dendograms will still be output.

```{r}
?heatmap.2

png(filename=paste0(projdir, "results/heatmap.png"))
heatmap <- heatmap.2(log2cpm_top50, 
                     dendrogram = "both", ## create a dendogram and reorder the rows by cluster
                     trace = "none", ## don't draw lines on top
                     scale = "row",  ## use z-score scaling for rows
                     col = bluered, ## use blues and reds
                     srtCol = 25, ## angle the column/sample labels since they are long
                     ColSideColors = col.grp, ## color sample groups as we did in the cluster plot (green ML, blue LP, red Basal)
                     margins = c(8,8)) 
dev.off()

# you can view the dendogram too although the gene name labels are hard to see
plot(heatmap$rowDendrogram, ylab="Cluster height", xlab="Gene")
plot(heatmap$colDendrogram, ylab="Cluster height", xlab="Sample")
```

This reflects what we saw earlier with the MDS and PCoA plots in that we can clearly see that the LP and MP groups are much more similar to one another than the Basal group. This is also doing some unsupervised clustering of both genes and samples, just with a different method of clustering (hierarchical). See the documentation for the `heatmap.2()` function for more information.

------------------------------------------------------------------------

##### Extract gene clusters from the heatmap

Sometimes you may want to pull out groups of interest from the heatmap. The `as.hclust` function can take the clustered data from the heatmap (the row dendogram that is, where genes are clustered).

```{r}
top50_clustered <- as.hclust(heatmap$rowDendrogram)
names(top50_clustered) ## these are the values stored - the height and order are included now - you can also view this in the environment tab
```

Now take this clustering information and "cut" the dendogram to select a fewer number of clusters to focus on. To do this, we need to cut based on the cluster height (higher numbers correspond to larger clusters). Look back at the dendogram above, or look at the distribution of height values. Picking a good cut point is somewhat of an art but often is practical (e.g. how many clusters do you want to keep?). Let's say for simplicity, we just want to know which genes fall into the clusters at height\>15 which should give us 4 gene clusters. We can use the `table` function to find out how many genes fall into each of the clusters. We can store these clusters in an object called `mycl`.

```{r}
?cutree
mycl <- cutree(top50_clustered, h=15)
table(mycl)
```

Now let's add the cluster ID to the rest of the data. The genes and cluster IDs should still be in the same order, but you can do a spot check first or to be really safe, you can join the data with the cluster IDs by SYMBOLs. We'll call this annotated version `cldat`.

```{r}
head(log2cpm_top50)
head(mycl)
cldat <- cbind(log2cpm_top50, clusterID=mycl)
head(cldat)
```

Notice the column called `clusterID` above. These are the cluster assigments for each gene (the numbers 1-4 are just identifiers - the values don't mean anything).

------------------------------------------------------------------------

##### Make volcano plots

Another way to visualize differential gene expression is via volcano plots. The `limma` package's function `volcanoplot()` function takes as input a fitted model object in the MArrayLM format (as is `efit`). We will create one for each of the pairwise comparisons that we did. I'll label the top 10 genes on each plot. If you want to remind yourself of the direction of the comparisons you can output the contrast matrix again.

```{r}
?volcanoplot
# basal vs lp
# png(filename=paste0(projdir, "results/basal_lp_volcano.png"))
volcanoplot(efit,
            coef = 1,
            style = "p-value",
            highlight = 10,
            names = names(efit$Amean),
            hl.col="blue",
            main = "Basal - LP",
            xlab = "Log2 Fold Change",
            ylab = NULL,
            pch=20,
            cex=0.35)
# dev.off()

# basal vs ml
# png(filename=paste0(projdir, "results/basal_ml_volcano.png"))
volcanoplot(efit, coef = 2, style = "B-statistic", highlight = 10, names = names(efit$Amean), 
            hl.col="blue",
            main = "Basal - ML", xlab = "Log2 Fold Change", ylab = NULL, pch=20, cex=0.35)
# dev.off()

# lp vs ml
# png(filename=paste0(projdir, "results/lp_ml_volcano.png"))
volcanoplot(efit,coef = 3, style = "p-value", highlight = 10, names = names(efit$Amean), 
            hl.col="blue", 
            main = "LP - ML", xlab = "Log2 Fold Change", ylab = NULL, pch=20, cex=0.35)
# dev.off()
```

This is a succinct way to visualize differential expression patterns, but let's pull out the labeled genes (top 10 most significant) for further observation.

```{r}
?topTable
# top hits:
BasalvLP_top <- topTable(efit, coef = 1, number = 10, sort.by = 'P')
BasalvLP_top

# top hits:
BasalvML_top <- topTable(efit, coef = 2, number = 10, sort.by = 'P')
BasalvML_top

# top hits:
ML_v_LP_top <- topTable(efit, coef = 3, number = 10, sort.by = 'P')
ML_v_LP_top
```

------------------------------------------------------------------------

##### Some notes on volcano plots

It is possible to refine this figure for publication, but I'd recommend using the `EnhancedVolcano` package to generate publication-ready plots.

In the interest of time, we won't be covering the EnhancedVolcano package for highly customizable plots. If you wish to give this a try yourself, see [this bioconductor vignette](https://bioconductor.org/packages/devel/bioc/vignettes/EnhancedVolcano/inst/doc/EnhancedVolcano.html). Fair warning: it will take some data cleaning and restructuring to implement.
